tests/avifgridapitest.c

```diff
@@ -0,0 +1,234 @@
+    for (int row = 0, iCell = 0; row < rows; ++row) {
+        for (int col = 0; col < columns; ++col, ++iCell) {
+            if (!createImage(cellWidth, cellHeight, depth, yuvFormat, &cellImages[iCell])) {
+                goto cleanup;
+            }
+        }
+    }
+
+    encoder = avifEncoderCreate();
+    if (encoder == NULL) {
+        printf("ERROR: avifEncoderCreate() failed\n");
+        goto cleanup;
+    }
+    encoder->speed = AVIF_SPEED_FASTEST;
+    if (avifEncoderAddImageGrid(encoder, columns, rows, (const avifImage * const *)cellImages, AVIF_ADD_IMAGE_FLAG_SINGLE) !=
+        AVIF_RESULT_OK) {
+        printf("ERROR: avifEncoderAddImageGrid() failed\n");
+        goto cleanup;
+    }
+    if (avifEncoderFinish(encoder, output) != AVIF_RESULT_OK) {
+        printf("ERROR: avifEncoderFinish() failed\n");
+        goto cleanup;
+    }
+
+    success = AVIF_TRUE;
+cleanup:
+    if (encoder != NULL) {
+        avifEncoderDestroy(encoder);
+    }
+    if (cellImages != NULL) {
+        for (int i = 0; i < columns * rows; ++i) {
+            if (cellImages[i] != NULL) {
+                avifImageDestroy(cellImages[i]);
+            }
+        }
+        avifFree(cellImages);
+    }
+    return success;
+}
+
+//------------------------------------------------------------------------------
+
+// Decodes the data. Returns false in case of failure.
+static avifBool decode(const avifRWData * encodedAvif)
+{
+    avifBool success = AVIF_FALSE;
+    avifImage * const image = avifImageCreateEmpty();
+    avifDecoder * const decoder = avifDecoderCreate();
+    if (image == NULL || decoder == NULL) {
+        printf("ERROR: memory allocation failed\n");
+        goto cleanup;
+    }
+    if (avifDecoderReadMemory(decoder, image, encodedAvif->data, encodedAvif->size) != AVIF_RESULT_OK) {
+        printf("ERROR: avifDecoderReadMemory() failed\n");
+        goto cleanup;
+    }
+    success = AVIF_TRUE;
+cleanup:
+    if (image != NULL) {
+        avifImageDestroy(image);
+    }
+    if (decoder != NULL) {
+        avifDecoderDestroy(decoder);
+    }
+    return success;
+}
+
+//------------------------------------------------------------------------------
+
+// Generates, encodes then decodes a grid image.
+static avifBool encodeDecode(int columns, int rows, int cellWidth, int cellHeight, avifPixelFormat yuvFormat)
+{
+    avifBool success = AVIF_FALSE;
+    avifRWData encodedAvif = { 0 };
+    if (!encodeGrid(columns, rows, cellWidth, cellHeight, /*depth=*/8, yuvFormat, &encodedAvif)) {
+        goto cleanup;
+    }
+    if (!decode(&encodedAvif)) {
+        goto cleanup;
+    }
+    success = AVIF_TRUE;
+cleanup:
+    avifRWDataFree(&encodedAvif);
+    return success;
+}
+
+//------------------------------------------------------------------------------
+
+// For each dimension, for each combination of cell count and size, generates, encodes then decodes a grid image.
+static avifBool encodeDecodeSizes(const int (*columnsCellWidths)[2],
```

Comment:
Nit/optional: The following syntax may be easier to understand:

const int columnsCellWidths[][2]